pkg/logging/logger.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2016-2019, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logging
+
+import (
+	"context"
+
+	"github.com/pulumi/pulumi/pkg/diag"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/provider"
+)
+
+// Logger wraps a time-ordered log set to allow batched logging of unique messages.
+type Logger struct {
+	messages *TimeOrderedLogSet
+	index    int
+}
+
+// NewLogger returns an initialized Logger.
+func NewLogger() *Logger {
+	return &Logger{
+		messages: &TimeOrderedLogSet{},
+	}
+}
+
+// EnqueueMessage adds a message to the log set.
+func (l *Logger) EnqueueMessage(severity diag.Severity, s string) {
+	l.messages.Add(Message{s, severity})
```

Comment:
Hmm, I think it's worth wondering whether we should be doing structured logging and de-dup'ing the structs rather than strings.

Reply:
I'm not sure I follow. What are you suggesting?

Reply:
```go
type LogMessage interface {
    Message() string
    Equal() bool
}
```

and then you have a bunch of message "types" implement this, instead of just representing them as strings.

The way it is now is fine, too.